Rewrite entire git procedure for release (#1544)
qutip/qutip: doc/development/release_distribution.rst

This document covers the process for managing updates to the current minor release and making new releases.
Within this document, the git remote ``upstream`` refers to the main QuTiP organsiation repository, and ``origin`` refers to your personal fork.

In short, the steps you need to take are:

1. Prepare the relevant ``git`` branch with the correct state of the code (git_).
2. Build the binary and source versions of the package, and deploy it to PyPI (``pip``) (deploy_).
3. Retrive the built documentation from GitHub (also in docbuild_).

> 4. Release the files on the QuTiP GitHub page (github_).

Reviewer: ```suggestion
4. Create a GitHub release and uploaded the built files to it (see github_).
```

5. Update `qutip.org <http://qutip.org/>`_ with the new links and documentation (web_).
6. Update the conda feedstock, deploying the package to ``conda`` (cforge_).

.. _git:

> Setting Up The Release Branch

Reviewer: I *think* one should not capitalize section headings, but I'm not enough of a grammar nazi to know for sure and I don't know if there is a QuTiP doc convention to follow.

Author: I'm pretty sure we don't have a convention.  Grammatically, title case (capitalise each word) is historically preferred (though you'd usually leave "the" without a capital), but in modern usage more and more people are just using sentence case.  I'm not especially bothered either way - I'm not even sure I'm consistent with myself day-to-day.

+++++++++++++++++++++++++++++

In this step you will prepare a git branch on the main QuTiP repository that has the state of the code that is going to be released.

> This procedure is quite different if you are releasing a new minor or major version compared to if you are making a bugfix patch release.

Reviewer: Is the procedure not the same except that if the `.X` branch doesn't exist you have to create it first?

Author: I see them as fairly different, aside from the changelog stuff, which I separated out to make it clear it's for both of them.  That said, it depends a little on the audience - to me and you I think "cherry-pick what you want onto a branch" is a very simple operation and not wildly dissimilar to "branch off and update a couple of files", but in the context of having to explain how to enumerate the commits and then how each one should be cherry-picked, it's really a very big difference between the two.   The files you need to update are different between the two as well, so there's a fair amount of difference.

Reviewer: Having read a bit further, I'm happy with having the two split.

#. Merge this into ``master``.

Now jump to release_ if you are making a major or minor release, or bugfix_ if you are only fixing bugs in a previous release.

.. _release:

Create a new minor or major release
-----------------------------------

This involves making a new branch to hold the release and adding some commits to set the code into "release" mode.
This release should be done by branching directly off the ``master`` branch at its current head.

#. On your machine, make sure your copy of ``master`` is up-to-date (``git checkout master; git pull upstream master``).
   This should at least involve fetching the changelog PR that you just made.
   Now create a new branch off a commit in ``master`` that has the state of the code you want to release.
   The command is ``git checkout -b qutip-<major>.<minor>.X``, for example ``qutip-4.7.X``.
   This branch name will be public, and must follow this format.
#. Push the new branch (with no commits in it relative to ``master``) to the main ``qutip/qutip`` repository (``git push upstream qutip-4.7.X``).

>    This is the only situation in which it is ok to push to ``qutip/qutip`` without making a pull request; you cannot make a pull request until the branch has been created.

Reviewer: ```suggestion
```
I would just remove this sentence. It's more to read and I think it's clear that releasing is a special process anyway. I'm also fairly sure that this is not the *only* case that it is okay to do this (e.g. `dev.major` presumably was created this way too).

Author: I'm not sure, to be honest.  `dev.major` was a hugely special case, whereas doing a release should be around every two months, including patch releases. I think `dev.major` was the first time QuTiP ever had two principle branches for simultaneous development (as opposed to short-lived features), and we don't really anticipate ever needing to make a third branch like it, since the branch just represents the "next major" release.

I would like to leave in some allusion to the fact that the guide tells you to push a branch directly, and this shouldn't be standard procedure.

Reviewer: Happy with this -- don't feel super strongly about the point.

Author: I'll just change it to "one of the few" and delete everything after the semicolon - makes it shorter, but still draws attention to it.

> #. Create a second new branch, which will be pushed to your fork and used to make a pull request against the ``qutip/qutip`` branch you just created.

Reviewer: ```suggestion
#. Create a second new branch, which will be pushed to your fork and used to make a pull request against the ``qutip-<major>.<minor>.X`` branch ``qutip/qutip`` you just created.
```

Author: I think you meant to add another word here maybe?  It doesn't seem to make sense.

Reviewer: Oops. I meant to write "branch on qutip/qutip" and not just "branch qutip/qutip". I intention was to make it clearer which branch was meant.

Author: Ok, that sounds better, and good plan to make the branch name clear as well.

   You will need to change the "base branch" in the pull request, because GitHub will always try to make the PR against ``master`` at first.
   When the tests have passed, merge this in.
#. Finally, back on ``master``, make a new pull request that changes the ``VERSION`` file to be ``<next-expected-version>.dev``, for example ``4.8.0.dev``.
   The "Development Status" in ``setup.cfg`` on ``master`` should not have changed, and should be ::

       Development Status :: 2 - Pre-Alpha

   because ``master`` is never directly released.

You should now have a branch that you can see on the GitHub website that is called ``qutip-4.7.X`` (or whatever minor version), and the state of the code in it should be exactly what you want to release as the new minor release.
If you notice you have made a mistake, you can make additional pull requests to the release branch to fix it.
``master`` should look pretty similar, except the ``VERSION`` will be higher and have a ``.dev`` suffix, and the "Development Status" in ``setup.cfg`` will be different.

You are now ready to actually perform the release.
Go to deploy_. 

.. _bugfix:

> Create a bug fix release

Reviewer: And here we have a lower case section heading, proving your point @jakelishman. :see_no_evil:

Author: Yep, I'm not even consistent with myself...  I think a very small majority of the section headings in the guide are in title case (capitalise most words), so let's go with that?

     Note the first 7 characters of its hash.
     Cherry-pick this by ``git cherry-pick --mainline 1 <hash>``.
   - "squash and merge": there will only be a single commit for the entire PR.
     Its name will be "<Name of the pull request> (#1000)".
     Note the first 7 characters of its hash.
     Cherry-pick this by ``git cherry-pick <hash>``.
   - "rebase and merge": this is the most difficult, because there will be many commits that you will have to find manually, and cherry-pick all of them.
     Go to the GitHub page for this PR, and go to the "Commits" tab.
     Using your local ``git log`` (you may find ``git log --oneline`` useful), find the hash for every single commit that is listed on the GitHub page, in order from **oldest to newest** (top-to-bottom in the GitHub view, which is bottom-to-top in ``git log``).
     You will need to use the commit message to do this; the hashes that GitHub reports will probably not be the same as how they appear locally.
     Find the first 7 characters of each of the hashes.

>      Cherry-pick these all in one go by ``git cherry-pick <hash1> <hash2> ... <hash10>``, where ``<hash1>`` is the oldest.

Reviewer: Nice description of all the cases here! :D
